Add Function mocking documentation (#279)
ballerina-platform/ballerina-dev-website: v1-2/learn/how-to-test-ballerina-code.md

* Test **assertions** can be used to verify the set of program behaviour expectations 
* Data providers can be used to feed in the test data sets 
* Service calls can be tested using service skeletons in the test phase of the project until the system is connected to the real service

> * Function mocks can be used to mimic third party function calls to enable testing a project module in isolation

Reviewer: ```suggestion
* Function mocks can be used to mimic third-party function calls to enable testing a project module in isolation.
```

Author: Changed as requested

    }
}
```

## Function Mocks

> Testerina provides the functionality to mock a function in a different thrird-part module with your own ballerina function, which will help you test your ballerina module independently.

Reviewer: ```suggestion
Testerina provides the functionality to mock a function in a different third-party module with your own Ballerina function, which will help you test your Ballerina module independently.
```

Author: Changed as requested

> Testerina provides the functionality to mock a function in a different thrird-party module with your own Ballerina function, which will help you test your Ballerina module independently.

Author: I have changed this to mention _"Testerina provides the functionality to mock a function from a differnt module in the project, a function from a 3rd party module or even a native Ballerina function"_

> Testerina provides the functionality to mock a function in a different thrird-party module with your own Ballerina function, which will help you test your Ballerina module independently.

Reviewer 2: How about "Ballerina test framework provides the capability to mock a function.  Using the mocking feature you can easily mock a function in a module that you are testing or a function of an imported module. This feature will help you to test your Ballerina code independently from other modules and functions"

> Testerina provides the functionality to mock a function in a different thrird-party module with your own Ballerina function, which will help you test your Ballerina module independently.

Author: Yes. This is much better. I will change the description to this

#### @test:Mock {}

> The function specified with the annotation `@test:Mock {}` will be considered as a mock function that gets triggered everytime the original function is called. The original function that will be mocked should be defined using the annotation parameters

Reviewer: ```suggestion
The function specified with the `@test:Mock {}` annotation will be considered as a mock function, which gets triggered every time the original function is called. The original function that will be mocked should be defined using the annotation parameters.
```

Author: Changed as requested

##### Annotation Value Fields

> `moduleName : "<moduleName>"` : Name of the module where the function to be mocked resides in. If the function is within the same module, this can be left blank or "." (No module) can be passed. If the function is in a different module, but within the same project, just passing the module name will suffice. For functions in completely seperate modules, the fully qualified module name must be passed, which includes the `orgName` and the `version`. ie. `orgName/module:version`. For native function, the ballerina module needs to be specified. 

Reviewer: ```suggestion
`moduleName : "<moduleName>"` : Name of the module where the function to be mocked resides in. If the function is within the same module, this can be left blank or "." (no module) can be passed. If the function is in a different module, but within the same project, just passing the module name will suffice. For functions in completely separate modules, the fully-qualified module name must be passed, which includes the `orgName` and the `version`. i.e., `orgName/module:version`. For native functions, the Ballerina module needs to be specified. 
```

Reviewer 2: Here the version is optional right ?

Author: When giving the fully qualified name, the version is also required in this case. Because it does all checks with the fully qualified module name.

`functionName : "<functionName>"` : Name of the function to be mocked.

Sample :

> The following is an example for function mocking

Reviewer: ```suggestion
The following is an example for function mocking.
```

Author: Changed as requested

```ballerina
import ballerina/io;
import ballerina/test;
import ballerina/math;

> // This is the mock function which will replace the real intAdd function

Reviewer: ```suggestion
// This is the mock function, which will replace the real `intAdd` function.
```

Author: Changed as requested

@test:Mock {

>     // Since the function is defined in the same module, we can pass "." as the current module

Reviewer: ```suggestion
    // Since the function is defined in the same module, "."  can be passed as the current module.
```

    // This can also be left blank.
    moduleName : ".",
    functionName : "intAdd"
}

> // The mock function signature should match the actual function signature

Reviewer: ```suggestion
// The mock function signature should match the actual function signature.
```

public function mockIntAdd(int a, int b) returns (int) {
    io:println("I am the mockIntAdd function");
    return (a - b);
}

> // This test function calls the local function intAdd() but it expects the mocked result

Reviewer: ```suggestion
// This test function calls the local `intAdd()`function but it expects the mocked result.
```

@test:Config {}  
function test_intAdd() {
    int answer = 0;
    answer = intAdd(5, 3);
    test:assertEquals(answer, 2, msg = "function mocking failed");
}

> public function intAdd(int a, int b) returns (int) {

Reviewer 2: better to split the sample in to actual files. ie intAdd in the module and mock in the test file. since we only support mocking in projects

Author: I split the function into a separate definition and I have mentioned that the function in the file is the one we are trying to mock

    return (a + b);
}

> // This test function calls the native function sqrt() but it expects the mocked result

Reviewer: ```suggestion
// This test function calls the native `sqrt()` function but it expects the mocked result.
```

@test:Config {}  
function test_sqrt() {
    float answer = 0;
    answer = math:sqrt(5);

    test:assertEquals(answer, 125.0, "mocking did not take place");
}

> // This is a mock function which will replace sqrt() from ballerina/math

Reviewer: ```suggestion
// This is a mock function, which will replace `sqrt()` by `ballerina/math`.
```
